modules/pipeline/pipengine/reconciler/queuemanage/manager/pipeline.go

```diff
@@ -0,0 +1,173 @@
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+package manager
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/modules/pipeline/pipengine/reconciler/queuemanage/queue"
+	"github.com/erda-project/erda/modules/pipeline/pipengine/reconciler/queuemanage/types"
+	"github.com/erda-project/erda/modules/pipeline/pipengine/reconciler/rlog"
+	"github.com/erda-project/erda/modules/pipeline/spec"
+	"github.com/erda-project/erda/pkg/loop"
+)
+
+// PutPipelineIntoQueue put pipeline into queue.
+// return: popCh, needRetryIfErr, err
+func (mgr *defaultManager) PutPipelineIntoQueue(pipelineID uint64) (<-chan struct{}, bool, error) {
+	// channel: send done signal when pipeline pop from the queue.
+	popCh := make(chan struct{})
+
+	// query pipeline detail
+	p := mgr.EnsureQueryPipelineDetail(pipelineID)
+	if p == nil {
+		return nil, false, fmt.Errorf("pipeline not found, pipelineID: %d", pipelineID)
+	}
+
+	// query pipeline queue detail
+	pq := mgr.EnsureQueryPipelineQueueDetail(pipelineID)
+	if pq == nil {
+		// pipeline doesn't bind queue, can reconcile directly
+		go func() {
+			popCh <- struct{}{}
+			close(popCh)
+		}()
+		return popCh, false, nil
+	}
+
+	// add queue to manager
+	q := mgr.addQueueToManager(pq)
+
+	// add pipeline to queue
+	q.AddPipelineIntoQueue(p, popCh)
+
+	// return channel when pipeline pop from queue
+	return popCh, false, nil
+}
+
+// addQueueToManager add one queue to mgr.
+func (mgr *defaultManager) addQueueToManager(pq *apistructs.PipelineQueue) types.Queue {
+	mgr.qLock.Lock()
+	defer mgr.qLock.Unlock()
+
+	// construct newQueue first for later use
+	newQueue := queue.New(pq, queue.WithDBClient(mgr.dbClient))
+
+	existQueue, ok := mgr.queueByID[newQueue.ID()]
+	if ok {
+		// update queue
+		existQueue.Update(pq)
+		mgr.queueByID[newQueue.ID()] = existQueue
+		return existQueue
+	}
+
+	// not exist, add new queue
+	mgr.queueByID[newQueue.ID()] = newQueue
+
+	return newQueue
+}
+
+// EnsureQueryPipelineDetail handle err properly.
+// return: pipeline or nil
+func (mgr *defaultManager) EnsureQueryPipelineDetail(pipelineID uint64) *spec.Pipeline {
+	mgr.pCacheLock.Lock()
+	defer mgr.pCacheLock.Unlock()
+
+	// try to get from pCache
+	cachedP, ok := mgr.pipelineCaches[pipelineID]
+	if ok {
+		return cachedP
+	}
+
+	// query from db
+	var p *spec.Pipeline
+	_ = loop.New(loop.WithDeclineLimit(time.Second*10), loop.WithDeclineRatio(2)).Do(func() (abort bool, err error) {
+		_p, exist, err := mgr.dbClient.GetPipelineWithExistInfo(pipelineID)
+		if err != nil {
+			err = fmt.Errorf("failed to query pipeline: %d, err: %v", pipelineID, err)
+			logrus.Error(err)
```

Comment:
logrus.Errorf()  Is it better

Reply:
you are right. I removed some logic here which use pre-defined err, so left it alone.
i will polish it.